File: js/src/http.ts
import type { ResolvedConfig } from "./config.js";
import { errorFromStatus, NetworkError } from "./errors.js";

export type Query = Record<string, string | number | boolean | undefined>;

export interface RequestOptions {
  body?: unknown;
  query?: Query;
}

export class Transport {
  constructor(
    private readonly cfg: ResolvedConfig,
    private readonly fetchImpl: typeof fetch = fetch,
  ) {}

  private authHeader(): string {
    const raw = `${this.cfg.clientId}:${this.cfg.apiKey}`;
    const b64 =
      typeof Buffer !== "undefined"
        ? Buffer.from(raw).toString("base64")
        : btoa(raw);
    return `Basic ${b64}`;
  }

  private buildUrl(path: string, query?: Query): string {
    let url = this.cfg.baseUrl + path;
    if (query) {
      const params = new URLSearchParams();
      for (const [k, v] of Object.entries(query)) {
        if (v !== undefined) params.append(k, String(v));
      }
      const qs = params.toString();
      if (qs) url += `?${qs}`;
    }
    return url;
  }

  async request<T = unknown>(
    method: string,
    path: string,
    opts: RequestOptions = {},
  ): Promise<T> {
    const url = this.buildUrl(path, opts.query);
    const headers: Record<string, string> = {
      authorization: this.authHeader(),
      accept: "application/json",
    };
    const init: RequestInit = { method, headers };
    if (opts.body !== undefined) {
      headers["content-type"] = "application/json";
      init.body = JSON.stringify(opts.body);
    }

    const controller = new AbortController();
    const timer = setTimeout(() => controller.abort(), this.cfg.timeoutMs);
    init.signal = controller.signal;

    let resp: Response;
    try {
      resp = await this.fetchImpl(url, init);
    } catch (err) {
      throw new NetworkError(`Request to ${url} failed`, err);
    } finally {
      clearTimeout(timer);
    }

    const text = await resp.text();
    const data = text ? JSON.parse(text) : undefined;

    if (!resp.ok) {
      const message =
        (data && typeof data === "object" && "message" in data
          ? String((data as Record<string, unknown>).message)
          : undefined) ?? `HTTP ${resp.status}`;
      let retryAfter: number | undefined;
      if (resp.status === 429) {
        const header = resp.headers.get("retry-after");
        if (header !== null) {
          const parsed = Number.parseInt(header, 10);
          if (Number.isFinite(parsed)) retryAfter = parsed;
        }
      }
      throw errorFromStatus(resp.status, message, data, retryAfter);
    }
    return data as T;
  }
}

File: python/callisto_sdk/_http.py
from __future__ import annotations

from typing import Any, Optional

import httpx

from ._config import Config
from ._reporter import ErrorReporter
from .errors import error_from_status, NetworkError


class Transport:
    def __init__(
        self,
        config: Config,
        http_client: Optional[httpx.Client] = None,
        reporter: Optional[ErrorReporter] = None,
    ):
        self._config = config
        self.reporter = reporter
        self._client = http_client or httpx.Client(
            auth=(config.client_id, config.api_key),
            timeout=config.timeout,
            headers={"accept": "application/json"},
        )

    def request(
        self,
        method: str,
        path: str,
        body: Optional[Any] = None,
        query: Optional[dict] = None,
    ) -> Any:
        url = self._config.base_url + path
        params = {k: v for k, v in (query or {}).items() if v is not None}
        try:
            resp = self._client.request(
                method, url, json=body if body is not None else None, params=params or None
            )
        except httpx.HTTPError as exc:
            err = NetworkError(f"Request to {url} failed: {exc}")
            self._report(err, method, path)
            raise err from exc

        data: Any = None
        if resp.content:
            try:
                data = resp.json()
            except ValueError:
                data = resp.text

        if resp.is_error:
            message = (
                data.get("message")
                if isinstance(data, dict) and "message" in data
                else f"HTTP {resp.status_code}"
            )
            retry_after: Optional[int] = None
            if resp.status_code == 429:
                raw = resp.headers.get("Retry-After")
                if raw is not None:
                    try:
                        retry_after = int(raw)
                    except ValueError:
                        retry_after = None
            err = error_from_status(resp.status_code, str(message), data, retry_after)
            self._report(err, method, path)
            raise err
        return data

    def _report(self, err: Any, method: str, path: str) -> None:
        if self.reporter is not None:
            self.reporter.capture_exception(err, method=method, path=path)

    def close(self) -> None:
        self._client.close()
